src/Bicep.Core/Semantics/Namespaces/NamespaceResolver.cs

```diff
@@ -0,0 +1,114 @@
+            {
+                var symbol = new BuiltInNamespaceSymbol(AzNamespaceType.BuiltInName, azNamespaceType);
+                builtInNamespaceSymbols[symbol.Name] = symbol;
+                namespaceTypes = namespaceTypes.Add(symbol.Name, azNamespaceType);
+            }
+
+            return new(namespaceTypes, builtInNamespaceSymbols.ToImmutableDictionary(LanguageConstants.IdentifierComparer));
+        }
+
+        public ImmutableDictionary<string, BuiltInNamespaceSymbol> BuiltIns { get; }
+
+        public IEnumerable<Symbol> ResolveUnqualifiedFunction(IdentifierSyntax identifierSyntax, bool includeDecorators)
+        {
+            // attempt to find function in all imported namespaces
+            var symbols = this.namespaceTypes.Values
+                .Select(type => includeDecorators
+                    ? type.MethodResolver.TryGetSymbol(identifierSyntax) ?? type.DecoratorResolver.TryGetSymbol(identifierSyntax)
+                    : type.MethodResolver.TryGetSymbol(identifierSyntax));
+
+            foreach (var symbol in symbols)
+            {
+                // The caller is responsible for deduplicating this list and returning an error if there is ambiguity, so we simply return everything.
+                if (symbol is not null)
+                {
+                    yield return symbol;
+                }
+            }
+        }
+
+        public IEnumerable<string> GetKnownFunctionNames(bool includeDecorators)
+            => this.namespaceTypes.Values
+                .SelectMany(type => includeDecorators
+                    ? type.MethodResolver.GetKnownFunctions().Keys.Concat(type.DecoratorResolver.GetKnownDecoratorFunctions().Keys)
+                    : type.MethodResolver.GetKnownFunctions().Keys);
+
+        public IEnumerable<string> GetKnownPropertyNames()
+            => this.namespaceTypes.Values.SelectMany(type => type.Properties.Keys);
+
+        public IEnumerable<string> GetNamespaceNames()
+            => this.namespaceTypes.Keys;
+
+        public ResourceType GetResourceType(ResourceTypeReference reference, ResourceTypeGenerationFlags flags)
+        {
+            var definedTypes = namespaceTypes.Values
+                .Select(type => type.ResourceTypeProvider.TryGetDefinedType(reference, flags))
+                .WhereNotNull();
+
+            if (definedTypes.FirstOrDefault() is {} definedType)
+            {
+                return definedType;
+            }
+
+            var generatedTypes = namespaceTypes.Values
+                .Select(type => type.ResourceTypeProvider.TryGenerateDefaultType(reference, flags))
+                .WhereNotNull();
+
+            // Here we are assuming that one of the namespaces will always return at least one result with TryGenerateDefaultType.
+            // This is a fair assumption at present, because the "az" namespace is always imported.
```

Comment:
>fair assumption [](http://example.com/codeflow?start=25&length=15)

Should the fallback/default type be part of the namespace contract? Or are you leaving it for later?

Reply:
It is part of the namespace contract; here I'm making the assumption that the `az` namespace provider will always return a fallback type (which is true as of now, but may not necessarily be in the future if we allow people to remove the `az` namespace).